### PR TITLE
#159 Improvements to python function deployment feature

### PR DESCRIPTION
**Pull request by nishadsingh1**
Corresponding issue: #159 

This PR modifies the python function deployment feature.

On the container, instead of recreating a new conda environment with `conda env create ...`, we now simply install pip and conda dependencies with `conda install` and `pip install` into the root conda environment.

On the client, we now write all temp files (exported environment files, new conda and pip dependency files) to /tmp instead of the user's working directory.

On the client, when parsing dependencies in the `environment.yml` (the exported conda environment file), we now remove any channel specifications accompanying packages. For instance, `conda-forge::package_name` in an `environment.yml` now gets reduced to `package_name` before being tested for package conflicts/existence. Including channel specifications limited the space in which we could search for package existence on the linux conda library, resulting in us skipping many packages that were actually available.

### CODE REVIEW

**Comment by nishadsingh1** [comment]
Because this PR makes changes to the Python function container image, we need to update our dockerhub image (https://hub.docker.com/r/clipper/python-container/) along with merging.

**Comment by dcrankshaw** [comment]
This is definitely an improvement, but it unfortunately didn't reduce the container build time. Here's the output from building the container:

```
Fetching package metadata .........
Solving package specifications: .

Package plan for installation in environment /opt/conda:

The following packages will be UPDATED:

    path.py:    10.0-py27_0       --> 10.3.1-py27_0
    pyasn1:     0.1.9-py27_0      --> 0.2.3-py27_0

The following packages will be DOWNGRADED due to dependency conflicts:

    cairo:      1.14.8-0          --> 1.12.18-6
    fontconfig: 2.12.1-2          --> 2.11.1-6
    harfbuzz:   0.9.39-2          --> 0.9.39-1
    jpeg:       9b-0              --> 8d-2
    libtiff:    4.0.6-3           --> 4.0.6-2
    matplotlib: 2.0.2-np112py27_0 --> 2.0.1-np112py27_0
    openssl:    1.0.2k-2          --> 1.0.2k-1
    pillow:     4.0.0-py27_0      --> 3.4.2-py27_0
    pixman:     0.34.0-0          --> 0.32.6-0
    qt:         5.6.2-4           --> 5.6.2-0
    requests:   2.14.2-py27_0     --> 2.13.0-py27_0

jpeg-8d-2.tar. 100% |###############################| Time: 0:00:00  15.76 MB/s
openssl-1.0.2k 100% |###############################| Time: 0:00:00  26.67 MB/s
pixman-0.32.6- 100% |###############################| Time: 0:00:00  28.20 MB/s
libtiff-4.0.6- 100% |###############################| Time: 0:00:00  26.41 MB/s
path.py-10.3.1 100% |###############################| Time: 0:00:00  35.76 MB/s
pyasn1-0.2.3-p 100% |###############################| Time: 0:00:00  43.68 MB/s
requests-2.13. 100% |###############################| Time: 0:00:00  34.97 MB/s
fontconfig-2.1 100% |###############################| Time: 0:00:00  28.86 MB/s
pillow-3.4.2-p 100% |###############################| Time: 0:00:00  23.21 MB/s
cairo-1.12.18- 100% |###############################| Time: 0:00:00  25.60 MB/s
harfbuzz-0.9.3 100% |###############################| Time: 0:00:00  24.45 MB/s
qt-5.6.2-0.tar 100% |###############################| Time: 0:00:01  27.37 MB/s
matplotlib-2.0  90% |###########################    | Time: 0:00:00  23.81 MB/s
matplotlib-2.0 100% |###############################| Time: 0:00:00  24.38 MB/s
Requirement already satisfied: backports-abc==0.5 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 1))
Requirement already satisfied: backports.shutil-get-terminal-size==1.0.0 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 2))
Requirement already satisfied: backports.ssl-match-hostname==3.4.0.2 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 3))
Requirement already satisfied: ipython-genutils==0.2.0 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 4))
Requirement already satisfied: jupyter-client==5.0.1 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 5))
Requirement already satisfied: jupyter-console==5.1.0 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 6))
Requirement already satisfied: jupyter-core==4.3.0 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 7))
Requirement already satisfied: prompt-toolkit==1.0.14 in /opt/conda/lib/python2.7/site-packages (from -r /model/pip_dependencies.txt (line 8))
Requirement already satisfied: six>=1.9.0 in /opt/conda/lib/python2.7/site-packages (from prompt-toolkit==1.0.14->-r /model/pip_dependencies.txt (line 8))
Requirement already satisfied: wcwidth in /opt/conda/lib/python2.7/site-packages (from prompt-toolkit==1.0.14->-r /model/pip_dependencies.txt (line 8))
Starting PythonContainer container
Connecting to Clipper with default port: 7000
Initializing Python function container
Serving predictions for doubles input type.
```

**Comment by rmdort** [comment]
Is conda a required dependency for all python containers ? `RPCDockerfile` is using `continuumio/anaconda` image. 

Does `pyzmq` or `rpc.py` depend on conda?

File: containers/python/python_container_entry.sh [hunk 1]
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env sh
 
-conda env create -f=/model/environment.yml -n container-env
-/bin/bash -c "source activate container-env && exec python /container/python_container.py"
+conda install -y --file /model/conda_dependencies.txt
+pip install -r /model/pip_dependencies.txt
+/bin/bash -c "exec python /container/python_container.py"
```
Comment (dcrankshaw) [inline containers/python/python_container_entry.sh]: So I had another idea about how to speed things up. Can you first try running the container without installing any of the supplied dependencies. If that fails, then install the dependencies and try running it again. So change `python_container_entry.sh` to something like:

```sh
/bin/bash -c "exec python /container/python_container.py
# if error...
conda install -y --file /model/conda_dependencies.txt
pip install -r /model/pip_dependencies.txt
/bin/bash -c "exec python /container/python_container.py"
```

Comment (nishadsingh1) [reply containers/python/python_container_entry.sh]: Oh good idea -- will give that a go

**Comment by dcrankshaw** [comment]
@rmdort You don't need to use Anaconda as your Python installation to write your own Docker containers. But we think Anaconda is pretty good overall and so we use it for our testing. The only part of Clipper that assumes Anaconda is the [`clipper_manager.Clipper.deploy_predict_function`](https://github.com/ucbrise/clipper/blob/develop/management/clipper_manager.py#L536) method.

**Comment by dcrankshaw** [comment]
LGTM. Rebased and formatted code. Will merge as soon as tests pass.

**Review by dcrankshaw — APPROVED** [review]
Brilliant. This worked great for simple stuff. We can fine tune the list of pre-installed packages as we go.

**Pull request merged**